crates/ruma-common/src/events/room/message/relation_serde.rs

```diff
@@ -66,133 +67,157 @@ where
-                    ..Default::default()
-                })
-            }
-            Relation::Replacement(Replacement { event_id, new_content }) => {
-                EventWithRelatesToJsonRepr {
-                    relates_to: RelatesToJsonRepr {
-                        relation: Some(RelationJsonRepr::Replacement(ReplacementJsonRepr {
-                            event_id: event_id.clone(),
-                        })),
-                        ..Default::default()
-                    },
-                    new_content: Some(new_content.clone()),
-                }
-            }
-            Relation::Thread(Thread { event_id, in_reply_to, is_falling_back }) => {
-                EventWithRelatesToJsonRepr::new(RelatesToJsonRepr {
-                    in_reply_to: in_reply_to.clone(),
-                    relation: Some(RelationJsonRepr::ThreadStable(ThreadStableJsonRepr {
-                        event_id: event_id.clone(),
-                        is_falling_back: *is_falling_back,
-                    })),
-                })
-            }
-            Relation::_Custom => EventWithRelatesToJsonRepr::<C>::default(),
-        };
+        let (relates_to, new_content) = self.clone().into_parts();
 
-        json_repr.serialize(serializer)
+        EventWithRelatesToSerHelper { relates_to, new_content }.serialize(serializer)
     }
 }
 
-#[derive(Deserialize, Serialize)]
-struct EventWithRelatesToJsonRepr<C> {
-    #[serde(rename = "m.relates_to", default, skip_serializing_if = "RelatesToJsonRepr::is_empty")]
-    relates_to: RelatesToJsonRepr,
+#[derive(Deserialize)]
+struct EventWithRelatesToDeHelper<C> {
+    #[serde(rename = "m.relates_to")]
+    relates_to: Option<RelatesToDeHelper>,
 
-    #[serde(rename = "m.new_content", skip_serializing_if = "Option::is_none")]
+    #[serde(rename = "m.new_content")]
     new_content: Option<C>,
 }
 
-impl<C> EventWithRelatesToJsonRepr<C> {
-    fn new(relates_to: RelatesToJsonRepr) -> Self {
-        Self { relates_to, new_content: None }
-    }
-}
-
-impl<C> Default for EventWithRelatesToJsonRepr<C> {
-    fn default() -> Self {
-        Self { relates_to: RelatesToJsonRepr::default(), new_content: None }
-    }
-}
-
-/// Struct modeling the different ways relationships can be expressed in a `m.relates_to` field of
-/// an event.
-#[derive(Default, Deserialize, Serialize)]
-struct RelatesToJsonRepr {
-    #[serde(rename = "m.in_reply_to", skip_serializing_if = "Option::is_none")]
+#[derive(Deserialize)]
+struct RelatesToDeHelper {
+    #[serde(rename = "m.in_reply_to")]
     in_reply_to: Option<InReplyTo>,
 
-    #[serde(flatten, skip_serializing_if = "Option::is_none")]
-    relation: Option<RelationJsonRepr>,
+    #[serde(flatten)]
+    relation: Option<RelationDeHelper>,
 }
 
-impl RelatesToJsonRepr {
-    fn is_empty(&self) -> bool {
-        self.in_reply_to.is_none() && self.relation.is_none()
-    }
+#[derive(Deserialize)]
+#[serde(untagged)]
+enum RelationDeHelper {
```

Comment:
Hm, this isn't much better than `serde_json::Value` I think (does internal buffering). I have a different idea..